components/containerd/pod/component.go

```diff
@@ -100,95 +149,170 @@ func (c *component) Metrics(ctx context.Context, since time.Time) ([]components.
+func parseUnixEndpoint(endpoint string) (string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", err
+	}
+	if u.Scheme != "unix" {
+		return "", fmt.Errorf("invalid scheme: %s (only supports 'unix' protocol)", u.Scheme)
+	}
+	return u.Path, nil
 }
 
-func (o *Output) States() ([]components.State, error) {
-	if len(o.Pods) == 0 {
-		return []components.State{{
-			Name:    containerd_pod_id.Name,
-			Health:  components.StateHealthy,
-			Healthy: true,
-			Reason:  "no output",
-		}}, nil
+// createConn creates a new CRI service client.
+// Cannot use "k8s.io/kubernetes/pkg/kubelet/cri/remote.NewRemoteRuntimeService" directly
+// as it causes a bunch of go module errors, importing the whole kubernetes repo.
+// ref. https://github.com/kubernetes-sigs/cri-tools/blob/master/cmd/main.go
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.29.2/pkg/kubelet/cri/remote/remote_runtime.go
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+func createConn(ctx context.Context, endpoint string) (runtimeapi.RuntimeServiceClient, runtimeapi.ImageServiceClient, *grpc.ClientConn, error) {
+	// "k8s.io/cri-client/pkg/util.GetAddressAndDialer" doesn't work...
+	// "code = Unavailable desc = name resolver error: produced zero addresses"
+	addr, err := parseUnixEndpoint(endpoint)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// "WithBlock" ctx cancel is no-op
+	conn, err := grpc.DialContext(ctx, addr, defaultDialOptions()...) //nolint:staticcheck
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	log.Logger.Debugw("successfully dialed -- checking version", "endpoint", endpoint)
+
+	// ref. https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+	runtimeClient := runtimeapi.NewRuntimeServiceClient(conn)
+	version, err := runtimeClient.Version(ctx, &runtimeapi.VersionRequest{})
+	if err != nil {
+		conn.Close()
+		return nil, nil, nil, err
 	}
+	log.Logger.Debugw("successfully checked version", "endpoint", endpoint, "version", version.String())
 
-	b, _ := o.JSON()
-	return []components.State{{
-		Name:    containerd_pod_id.Name,
-		Health:  components.StateHealthy,
-		Healthy: true,
-		Reason:  o.describeReason(),
-		ExtraInfo: map[string]string{
-			"data":     string(b),
-			"encoding": "json",
-		},
-	}}, nil
+	status, err := runtimeClient.Status(ctx, &runtimeapi.StatusRequest{})
+	if err != nil {
+		conn.Close()
+		return nil, nil, nil, err
+	}
+	log.Logger.Debugw("successfully checked status", "endpoint", endpoint, "status", status.String())
+
+	imageClient := runtimeapi.NewImageServiceClient(conn)
+	return runtimeClient, imageClient, conn, nil
 }
 
-const (
-	defaultSocketFile               = "/run/containerd/containerd.sock"
-	defaultContainerRuntimeEndpoint = "unix:///run/containerd/containerd.sock"
-)
+func CheckContainerdRunning(ctx context.Context) bool {
```

Comment:
Actually we don't need this function because the component should be enabled by default. and we should add a flag to control which components can be disabled or enabled(for those components disabled by default).